src/main/java/frc/team2412/robot/subsystem/TargetLocalizer.java

```diff
@@ -0,0 +1,107 @@
+package frc.team2412.robot.subsystem;
+
+import frc.team2412.robot.Robot;
+import org.frcteam2910.common.math.Rotation2;
+
+import static frc.team2412.robot.subsystem.TargetLocalizer.LocalizerConstants.*;
+
+public class TargetLocalizer {
+    public static class LocalizerConstants {
+        public static final double TURRET_OFFSET = 0;
+        // TODO tune these more
+        public static final double TURRET_LATERAL_FF = 0.1, TURRET_ANGULAR_FF = 5, TURRET_DEPTH_FF = 0.1;
+    }
+
+    private final DrivebaseSubsystem drivebaseSubsystem;
+    private final ShooterSubsystem shooterSubsystem;
+    private final ShooterVisionSubsystem shooterVisionSubsystem;
+
+    public TargetLocalizer(DrivebaseSubsystem drivebase, ShooterSubsystem shooter, ShooterVisionSubsystem vision) {
+        drivebaseSubsystem = drivebase;
+        shooterSubsystem = shooter;
+        shooterVisionSubsystem = vision;
+    }
+
+    public double getDistance() {
+        return hasTarget() ? shooterVisionSubsystem.getDistance() + shooterSubsystem.getDistanceBias() : 0;
+    }
+
+    public double getAdjustedDistance() {
+        return getDistance() + distanceAdjustment();
+    }
+
+    /**
+     * very basic feedforward math to adjust the depth depending on the distance you are moving away
+     * from target
+     *
+     * @return adjustment
+     */
+    public double distanceAdjustment() {
+        return (getDepthVelocity() * getDistance() * TURRET_DEPTH_FF) / getVoltage();
+    }
+
+    public double getPitch() {
+        return shooterVisionSubsystem.getAdjustedPitch();
+    }
+
+    public boolean hasTarget() {
+        return shooterVisionSubsystem.hasTarget();
+    }
+
+    public double getYaw() {
+        // return 0;
+        return shooterVisionSubsystem.getYaw() + shooterSubsystem.getTurretAngleBias();
+    }
+
+    /**
+     * unit vector component of chassis velocity perpendicular to the turret output
+     *
+     * @return that
+     */
+    public double getLateralVelocity() {
+        // might need to do inverse
+        return drivebaseSubsystem.getVelocity()
+                .rotateBy(Rotation2.fromDegrees(TURRET_OFFSET + shooterSubsystem.getTurretAngle())).x;
```

Comment:
I'm pretty sure it should be the component of the velocity perpendicular to (vector from the robot to) the hub, though the turret angle is probably a good enough approximation. (There might be some funny feedback stuff, though.)
Also, I think the angle we're rotating by should be inverted- If from the robot's point of view the robot is moving straight forward and the turret is to the right (clockwise), the lateral velocity component should be negative (assuming positive x is right and positive y is forward), which happens if we rotate to the left (counterclockwise). (Another way to think about it is that we need to rotate in the opposite direction to bring the turret's straight forward to the right spot.)

Reply:
(Splitting into separate comments for clarity)

Regarding the angle to rotate by - I'm not convinced that rotating the velocity by the inverted turret angle is the correct approach either. Consider the case where the robot is driving at a 45 degree angle (from its point of view) and the turret is pointed at a -45 degree angle (so the turret is at a right angle to the direction the robot is moving). We'd want the lateral velocity component to be negative so the turret turns counterclockwise as the robot moves. But if we rotate the velocity vector by 45 degrees, we'll have a vector at a 90 degree angle, which would have a positive x-component (lateral velocity). Rotating the vector by -45 degrees also wouldn't work, as this would have a lateral velocity of 0. (Here I'm assuming 0 degrees is in the +Y direction and rotating clockwise is positive, although I think the issue still persists with other conventions).

Reply:
I think the only problem with the 90 degree angle vector is that we should subtract the lateral velocity in the yaw adjustment instead of adding it.

Reply:
tuning will fix this

Reply:
Tuning wouldn't fix the case I mentioned - with a lateral velocity of zero, no adjustment would ever be made regardless of the feedforward value.

Reply:
it loads now wohoo

Reply:
we can swap some math when we tune.
also we can straight up delete this part if it has issues